File: py-core/dongnae/engine.py
"""
Key concept of this engine is "dongnae" - an object that has 'ID(dnid), Name(dnname), 2D coordinates(dnlatitude, dnlongitude), and radius(dnradius).
Dictionary of dongnaes should be loaded from CSV prior to using this engine.
"""

import csv
import math
from typing import List, Dict, Optional, Union

class DongnaeEngine:
    def __init__(self, csv_path: str = None):
        self._dongnaes: List[Dict] = []
        
        # Default Haversine coefficients (Based on Korea, approx 37N)
        # Will be updated automatically in load()
        self._lat_coef = 111.0
        self._lon_coef = 88.8
        
        # load CSV if path provided
        if csv_path:
            self.load(csv_path)

    def load(self, csv_path: str):
        """
        Loads the optimized CSV file into memory with automatic encoding detection.
        Also auto-calculates Haversine coefficients based on the dataset's latitude.
        """
        encodings = ['utf-8-sig', 'cp949', 'utf-8']
        loaded = False
        
        for enc in encodings:
            try:
                with open(csv_path, mode='r', encoding=enc) as f:
                    reader = csv.DictReader(f)
                    temp_data = []
                    for row in reader:
                        temp_data.append({
                            'dnid': row['dnid'],
                            'dnname': row['dnname'],
                            'dnlatitude': float(row['dnlatitude']),
                            'dnlongitude': float(row['dnlongitude']),
                            'dnradius': float(row['dnradius'])
                        })
                    self._dongnaes = temp_data
                    loaded = True
                    break

            except (UnicodeDecodeError, KeyError, ValueError):
                continue
        
        if not loaded:
            raise ValueError(f"Failed to load CSV: {csv_path}. Tried encodings: {encodings}.")

        # -------------------------------------------------------
        # [Auto-Calibration] Calculate Haversine coefficients
        # -------------------------------------------------------
        if self._dongnaes:
            # 1. Extract all latitudes
            lats = [d['dnlatitude'] for d in self._dongnaes]
            
            # 2. Find the center latitude of the dataset
            min_lat, max_lat = min(lats), max(lats)
            avg_lat = (min_lat + max_lat) / 2.0
            
            # 3. Update coefficients (Round to 2 decimal places)
            # Latitude: Approx 111 km per degree (Constant)
            # Longitude: 111 * cos(lat) km per degree (Varies by latitude)
            self._lat_coef = 111.0
            self._lon_coef = round(111.0 * math.cos(math.radians(avg_lat)), 2)

    def _calc_dist(self, lat1: float, lon1: float, lat2: float, lon2: float) -> float:
        """
        [Geometric Distance] Calculates distance between two points (Haversine approximation).
        Uses dynamically calculated coefficients based on the loaded dataset's center latitude.
        """
        d_lat = (lat2 - lat1) * self._lat_coef
        d_lon = (lon2 - lon1) * self._lon_coef
        return math.sqrt(d_lat**2 + d_lon**2)

    def _dongnae_dist(self, lat: float, lon: float, dn: Dict) -> float:
        """
        [Business Metric] Calculates 'Boundary Distance' from a point to a Dongnae.
        Returns: (Geometric Distance to Center) - (Radius of Dongnae)
        """
        # Delegate geometric calculation to _calc_dist
        center_dist = self._calc_dist(lat, lon, dn['dnlatitude'], dn['dnlongitude'])
        
        # Apply Radius adjustment
        return center_dist - dn['dnradius']

    def where(self, lat: float, lon: float) -> Optional[Dict]:
        """
        [Reverse Geocoding] Returns the single nearest 'Dongnae' (neighborhood).
        """
        nearest = self.nearest(lat, lon, k=1)
        return nearest[0] if nearest else None

    def nearest(self, lat: float, lon: float, k: int = 1, radius_km: float = None) -> List[Dict]:
        """
        Returns the K nearest Dongnaes sorted by 'Boundary Distance'.
        :param radius_km: Used to limit the search range (performance optimization)
        """
        # 1. Primary filtering (Bounding Box)
        # Search range = (Requested Radius OR Default 10km) + Max Dongnae Radius Buffer(5km)
        scan_radius = (radius_km if radius_km else 10.0) + 5.0
        
        # Dynamic bbox calculation using current coefficients
        lat_delta = scan_radius / self._lat_coef
        lon_delta = scan_radius / self._lon_coef
        
        candidates = [
            dn for dn in self._dongnaes
            if (lat - lat_delta <= dn['dnlatitude'] <= lat + lat_delta) and
               (lon - lon_delta <= dn['dnlongitude'] <= lon + lon_delta)
        ]
        
        if not candidates and radius_km is None:
            candidates = self._dongnaes

        # 2. Calculate Boundary Distance
        results = []
        for dn in candidates:
            b_dist = self._dongnae_dist(lat, lon, dn)
            
            if radius_km is None or b_dist <= radius_km:
                dn_res = dn.copy()
                dn_res['distance'] = round(b_dist, 4)
                results.append(dn_res)

        results.sort(key=lambda x: x['distance'])
        return results[:k]

    def within(self, lat: float, lon: float, radius_km: float, limit: int = None) -> List[Dict]:
        """
        [Radius Search] Returns all Dongnaes whose boundaries are within R km.
        """
        return self.nearest(lat, lon, k=limit if limit else len(self._dongnaes), radius_km=radius_km)

    def resolve(self, lat: float, lon: float, threshold: float = 1.0) -> List[Dict]:
        """
        [Soft Geofencing] Determine if coordinates fall within a specific Dongnae's effective radius.
        """
        max_scan = 15.0 * threshold
        lat_delta = max_scan / self._lat_coef
        lon_delta = max_scan / self._lon_coef

        candidates = [
            dn for dn in self._dongnaes
            if (lat - lat_delta <= dn['dnlatitude'] <= lat + lat_delta) and
               (lon - lon_delta <= dn['dnlongitude'] <= lon + lon_delta)
        ]

        matches = []
        for dn in candidates:
            b_dist = self._dongnae_dist(lat, lon, dn)
            
            if b_dist <= dn['dnradius'] * (threshold - 1.0):
                dn_res = dn.copy()
                dn_res['distance'] = round(b_dist, 4)
                
                raw_dist = b_dist + dn['dnradius']
                limit_dist = dn['dnradius'] * threshold
                
                dn_res['score'] = round(raw_dist / limit_dist, 2)
                matches.append(dn_res)
        
        matches.sort(key=lambda x: x['score'])
        return matches


    def search(self, keyword: str, limit: int = 5, best_shot: bool = True) -> Union[List[Dict], Optional[Dict]]:
        """
        [Text Search & Geocoding] Search by Dongnae name (Bag of Words similarity).
        
        :param keyword: Search query (e.g., "Pangyo")
        :param limit: Max number of candidates (only used when best_shot=False)
        :param best_shot: If True, returns the single best match (Geocoding mode).
                          If False, returns a list of candidates (Search mode).
        """
        query_tokens = keyword.strip().split()
        if not query_tokens:
            return None if best_shot else []

        scored_list = []
        for dn in self._dongnaes:
            score = 0
            name = dn['dnname']
            for token in query_tokens:
                if token in name:
                    score += 1
            
            if score > 0:
                scored_list.append({
                    'data': dn,
                    'score': score,
                    'len': len(name) # Prefer shorter names (tie-breaker)
                })
        
        # Sort: High score -> Short name length
        scored_list.sort(key=lambda x: (-x['score'], x['len']))
        
        # Inject score into results
        results = []
        # If best_shot is True, we only need the top 1, otherwise up to limit
        target_slice = scored_list[:1] if best_shot else scored_list[:limit]
        
        for item in target_slice:
            dn_res = item['data'].copy()
            dn_res['score'] = item['score']
            results.append(dn_res)
            
        if not results:
            return None if best_shot else []

        # Return Logic based on best_shot flag
        if best_shot:
            return results[0]  # Return single Dict (Geocoding Mode)
        else:
            return results     # Return List[Dict] (Search Mode)


    def get(self, dnid: str) -> Optional[Dict]:
        """
        [ID Lookup] Direct lookup by dnid (legal dong code)
        """
        for dn in self._dongnaes:
            if dn['dnid'] == str(dnid):
                return dn
        return None

File: js-data-kr/src/engine.js
import rawData from './dongnae_data.js';

export class DongnaeEngine {
    constructor() {
        // 1. 데이터 바인딩 (메모리 복사 없이 참조만 함)
        // SoA (Structure of Arrays) 패턴
        this.ids = rawData.ids;
        this.names = rawData.names;
        this.lats = rawData.lats;
        this.lons = rawData.lons;
        this.rads = rawData.rads;
        this.count = this.ids.length;

        // 2. 계수 자동 보정 (Auto-Calibration)
        // 위도 전체를 순회하여 min/max와 최대 반경을 구합니다.
        // (100개 샘플링은 진짜 극값을 놓쳐 Python 코어와 계수가 달라질 수 있으므로
        //  전수 스캔으로 통일. 초기화 1회뿐이라 비용은 무시할 수준입니다.)
        let minLat = 90, maxLat = -90, maxRad = 0;
        for (let i = 0; i < this.count; i++) {
            const val = this.lats[i];
            if (val < minLat) minLat = val;
            if (val > maxLat) maxLat = val;
            if (this.rads[i] > maxRad) maxRad = this.rads[i];
        }
        const avgLat = (minLat + maxLat) / 2.0;

        this.latCoef = 111.0;
        // JS의 Math.cos는 라디안을 받습니다.
        this.lonCoef = +(111.0 * Math.cos(avgLat * (Math.PI / 180))).toFixed(2);

        // 데이터셋 내 최대 반경. bbox 버퍼가 이 값을 따라가도록 하여
        // 큰 반경 노드가 후보에서 누락되지 않게 합니다. (Python 코어와 동일)
        this.maxRadius = maxRad;

        // 3. ID 검색용 인덱스 생성 (O(1) Lookup)
        this.idMap = new Map();
        for (let i = 0; i < this.count; i++) {
            this.idMap.set(this.ids[i], i);
        }
    }

    /**
     * [Internal] 특정 인덱스의 데이터를 객체로 변환하여 반환
     * 필요할 때만 객체를 생성하므로 메모리를 절약함 (Lazy Generation)
     */
    _getAt(i, distance = null, score = null) {
        const obj = {
            dnid: this.ids[i],
            dnname: this.names[i],
            dnlatitude: this.lats[i],
            dnlongitude: this.lons[i],
            dnradius: this.rads[i]
        };
        if (distance !== null) obj.distance = distance;
        if (score !== null) obj.score = score;
        return obj;
    }

    /**
     * [Internal] 거리 계산 (Haversine approximation)
     */
    _calcDist(lat1, lon1, lat2, lon2) {
        const dLat = (lat2 - lat1) * this.latCoef;
        const dLon = (lon2 - lon1) * this.lonCoef;
        return Math.sqrt(dLat * dLat + dLon * dLon);
    }

    /**
     * [Internal] 동네 경계 거리 계산
     * (중심점 거리) - (반경)
     */
    _dongnaeDist(lat, lon, idx) {
        const centerDist = this._calcDist(lat, lon, this.lats[idx], this.lons[idx]);
        return centerDist - this.rads[idx];
    }

    howfar(lat, lon, dnid) {
        // 1. ID 존재 여부 확인 (O(1) Lookup)
        // 입력된 dnid를 문자열로 안전하게 변환하여 조회
        const idx = this.idMap.get(String(dnid));

        // 존재하지 않는 ID인 경우 null 반환
        if (idx === undefined) {
            return null;
        }

        // 2. 내부 메서드를 재사용하여 경계 거리 계산
        // (중심점까지의 거리) - (동네 반경)
        const dist = this._dongnaeDist(lat, lon, idx);

        // raw 값 반환 (Python 코어 howfar과 동일하게 반올림하지 않음)
        return dist;
    }

    /**
     * [Reverse Geocoding] Lookup one nearest Dongnae from given coordinate
     */
    where(lat, lon) {
        const results = this.nearest(lat, lon, 1);
        return results.length > 0 ? results[0] : null;
    }

    /**
     * [Nearest Neighbor] Lookup k-best Dongnae(s) from given coordinate
     */
    nearest(lat, lon, k = 1, radiusKm = null) {
        // 1. Bounding Box Filtering
        // 검색 범위 + 데이터셋 최대 반경(버퍼). 큰 반경 노드도 후보에 포함되도록 보장.
        // (?? 사용: radiusKm=0도 유효한 0 범위로 취급, falsy 함정 방지)
        const effectiveRange = radiusKm ?? 10.0;
        const scanRadius = effectiveRange + this.maxRadius;
        const latDelta = scanRadius / this.latCoef;
        const lonDelta = scanRadius / this.lonCoef;

        const candidates = [];

        for (let i = 0; i < this.count; i++) {
            const tLat = this.lats[i];
            const tLon = this.lons[i];

            // check radius (BBox)
            if (tLat >= lat - latDelta && tLat <= lat + latDelta &&
                tLon >= lon - lonDelta && tLon <= lon + lonDelta) {

                const bDist = this._dongnaeDist(lat, lon, i);

                if (radiusKm === null || bDist <= radiusKm) {
                    candidates.push({ idx: i, dist: parseFloat(bDist.toFixed(4)) });
                }
            }
        }

        // Fallback parity with the Python core: if the bbox produced no
        // candidates and no radius limit was given, scan the whole dataset so
        // where()/nearest() stay total (never empty for a non-empty dataset).
        if (candidates.length === 0 && radiusKm === null) {
            for (let i = 0; i < this.count; i++) {
                const bDist = this._dongnaeDist(lat, lon, i);
                candidates.push({ idx: i, dist: parseFloat(bDist.toFixed(4)) });
            }
        }

        // 2. Sort by Distance
        candidates.sort((a, b) => a.dist - b.dist);

        // 3. Map to Objects
        return candidates.slice(0, k).map(c => this._getAt(c.idx, c.dist));
    }

    /**
     * [Radius Search] search Dongnae(s) within certain radius
     */
    within(lat, lon, radiusKm, limit = null) {
        const results = this.nearest(lat, lon, this.count, radiusKm);
        // limit != null: limit=0이면 빈 배열 (falsy 함정 방지)
        return limit != null ? results.slice(0, limit) : results;
    }

    /**
     * [Soft Geofencing] 특정 영역 포함 여부 판정
     */
    resolve(lat, lon, threshold = 1.0) {
        // 매칭 가능한 최대 중심거리 = maxRadius * threshold. (고정 15km 가정 제거)
        const maxScan = this.maxRadius * threshold;
        const latDelta = maxScan / this.latCoef;
        const lonDelta = maxScan / this.lonCoef;

        const matches = [];

        for (let i = 0; i < this.count; i++) {
            const tLat = this.lats[i];
            const tLon = this.lons[i];

            if (tLat >= lat - latDelta && tLat <= lat + latDelta &&
                tLon >= lon - lonDelta && tLon <= lon + lonDelta) {

                const bDist = this._dongnaeDist(lat, lon, i);
                const radius = this.rads[i];

                // 판정 로직
                if (bDist <= radius * (threshold - 1.0)) {
                    const rawDist = bDist + radius;
                    const limitDist = radius * threshold;
                    // limitDist 0 (radius/threshold 0) => 정확히 중심점, score 0.0
                    // (Python 코어와 동일, NaN 방지)
                    const score = limitDist ? parseFloat((rawDist / limitDist).toFixed(2)) : 0.0;

                    matches.push(this._getAt(i, parseFloat(bDist.toFixed(4)), score));
                }
            }
        }

        matches.sort((a, b) => a.score - b.score);
        return matches;
    }

    /**
     * [Text Search] 키워드 검색
     */
    search(keyword, limit = 5, bestShot = true) {
        // filter(Boolean): 빈 문자열/공백 입력 시 [""]가 되어 모든 이름에
        // 매칭되는 버그 방지. 토큰이 없으면 결과 없음으로 처리.
        const queryTokens = keyword.trim().split(/\s+/).filter(Boolean);
        if (queryTokens.length === 0) return bestShot ? null : [];

        const scoredList = [];

        for (let i = 0; i < this.count; i++) {
            let score = 0;
            const name = this.names[i];

            for (const token of queryTokens) {
                if (name.includes(token)) score++;
            }

            if (score > 0) {
                scoredList.push({
                    idx: i,
                    score: score,
                    len: name.length
                });
            }
        }

        // Sort: Score Desc -> Length Asc
        scoredList.sort((a, b) => {
            if (b.score !== a.score) return b.score - a.score;
            return a.len - b.len;
        });

        const targetSlice = bestShot ? scoredList.slice(0, 1) : scoredList.slice(0, limit);
        const results = targetSlice.map(item => this._getAt(item.idx, null, item.score));

        if (results.length === 0) return bestShot ? null : [];
        return bestShot ? results[0] : results;
    }

    /**
     * [ID Lookup] ID로 직접 조회 (O(1))
     */
    get(dnid) {
        const idx = this.idMap.get(String(dnid)); // 문자열로 안전하게 변환
        return idx !== undefined ? this._getAt(idx) : null;
    }
}